forte/data/data_store.py

```diff
@@ -493,6 +495,129 @@ def get_entry(self, tid: int) -> Tuple[List, int, int]:
         # Otherwise, use ``index_id`` to find the index of the entry.
         raise NotImplementedError
 
+    def co_iterator(self, type_names: List[str]) -> Iterator[List]:
+        r"""
+        Given two or more type names, iterate their entry lists from beginning to end together.
+
+        For every single type, their entry lists are sorted by the begin and
+        end fields. The co_iterator function will iterate those sorted lists
+        together, and yield each entry in sorted order. This tasks is quite
+        similar to merging several sorted list to one sorted list. We internally
+        use a `MinHeap` to order the order of yielded items, and the ordering
+        is determined by:
+
+            - start index of the entry.
+            - end index of the entry.
+            - the index of the entry type name in input parameter ``type_names``.
+
+        The precedence of those values indicates their priority in the min heap
+        ordering.
+        For example, if two entries have both the same begin and end field,
+        then their order is
+        decided by the order of user input type_name (the type that first
+        appears in the target type list will return first).
+
+        Args:
+            type_names (List[str]): a list of string type names
+
+        Returns:
+
+            An iterator of entry elements.
+
+        """
+
+        n = len(type_names)
+        # suppose the length of type_names is N and the length of entry list of
+        # one type is M
+        # then the time complexity of using min-heap to iterate
+        # is O(M*log(N))
+
+        # Initialize the first entry of all entry lists
+        # it avoids empty entry lists or non-existant entry list
+        first_entries = []
+        for tn in type_names:
+            if tn in self.__elements:
+                if len(self.__elements[tn]) > 0:
+                    first_entries.append(self.__elements[tn][0])
+                else:
```

Comment:
Here you can actually simplify a little bit by catching the `IndexError` to find out that the entry list is empty.